apps/aehttp/src/aehttp_dispatch_rosetta.erl

```diff
@@ -0,0 +1,659 @@
+        Class:Reason:Stacktrace ->
+            lager:error("CRASH ~p ~p, ~p", [Class, Reason, Stacktrace]),
+            {500, [], #{reason => <<"Internal server error">>}}
+    end.
+
+%% read transactions
+%% Data API - Network
+queue(networkList) ->
+    ?READ_Q;
+queue(networkOptions) ->
+    ?READ_Q;
+queue(networkStatus) ->
+    ?READ_Q;
+%% Data API - Account
+queue(accountBalance) ->
+    ?READ_Q;
+queue(accountCoins) ->
+    ?READ_Q;
+%% Data API - Block
+queue(block) ->
+    ?READ_Q;
+queue(blockTransaction) ->
+    ?READ_Q;
+%% Data API - Mempool
+queue(mempool) ->
+    ?READ_Q;
+queue(mempoolTransaction) ->
+    ?READ_Q;
+%% Construction API - Construction
+queue(constructionCombine) ->
+    ?WRITE_Q;
+queue(constructionDerive) ->
+    ?WRITE_Q;
+queue(constructionHash) ->
+    ?WRITE_Q;
+queue(constructionMetadata) ->
+    ?WRITE_Q;
+queue(constructionParse) ->
+    ?WRITE_Q;
+queue(constructionPayloads) ->
+    ?WRITE_Q;
+queue(constructionPreprocess) ->
+    ?WRITE_Q;
+queue(constructionSubmit) ->
+    ?WRITE_Q;
+%% Indexers
+queue(call) ->
+    ?READ_Q;
+queue(eventsBlocks) ->
+    ?READ_Q;
+queue(searchTransactions) ->
+    ?READ_Q;
+queue(_) ->
+    ?WRITE_Q.
+
+handle_request_(networkList, _, _Context) ->
+    Resp =
+        #{<<"network_identifiers">> =>
+              [#{<<"blockchain">> => aeapi:blockchain_name(),
+                 <<"network">> => aeapi:network_id()}]},
+    {200, [], Resp};
+handle_request_(networkOptions, _, _Context) ->
+    Resp =
+        #{<<"version">> =>
+              #{<<"rosetta_version">> => <<"1.4.10">>,
+                <<"node_version">> => aeapi:node_version(),
+                <<"metadata">> => #{<<"node_revision">> => aeapi:node_revision()}},
+          <<"allow">> =>
+              #{<<"operation_statuses">> =>
+                    [#{<<"status">> => <<"SUCCESS">>, <<"successful">> => true},
+                     #{<<"status">> => <<"FAILED">>, <<"successful">> => false}],
+                <<"operation_types">> => ae_operation_types(),
+                <<"errors">> => rosetta_errors(),
+                <<"historical_balance_lookup">> => true,
+                <<"timestamp_start_index">> => 1,
+                <<"call_methods">> => [<<"TODO">>],
+                <<"balance_exemptions">> => [],
+                <<"mempool_coins">> => false}},
+    {200, [], Resp};
+handle_request_(networkStatus, _, _Context) ->
+    try
+        {ok, TopBlock} = aeapi:top_key_block(),
+        CurrentBlock =
+            case aec_blocks:height(TopBlock) of
+                0 ->
+                    TopBlock;
+                _ ->
+                    {ok, Prev} = aeapi:prev_key_block(TopBlock),
+                    Prev
+            end,
+        CurrentBlockIdentifier = format_block_identifier(CurrentBlock),
+        CurrentBlockTimestamp = aec_blocks:time_in_msecs(CurrentBlock),
+        GenesisBlockIdentifier = format_block_identifier(aec_chain:genesis_block()),
```

Comment:
Should be done within one activity

Reply:
Perhaps consider whether activity wrappers should be placed uniformly at the top level, or whether some should be in helper functions. The `ensure_activity/2` semantics allows for a combination, but also consider what should be dirty and where transactions are needed. As a point of implementation, an outer `transaction` will convert inner `async_dirty` activities to re-use the outer transaction context.

Reply:
Nice suggestion. I'll look at this in the next phase of Rosetta